Add new skin layer to allow components anchored to the playfield

peppy opened the pull request:
This has been a common request to cater for rulesets where the playfield changes in shape (ie. osu!mania with different column counts), or for cases where a user has applied playfield scale adjustments.

See commit message attached to 5bd06832d0a4e35bbbd52eea2ffb3970764e6d0e for the major change required to make this work.


https://github.com/ppy/osu/assets/191335/19be5e65-e7fe-4cc0-9e16-44b0ce18a0ee


Addresses #24379.
Closes https://github.com/ppy/osu/issues/22388

peppy committed:
Expose a new SSDQ from playfield for skinnable area bounds

peppy committed:
Add playfield layer to skin editor

peppy committed:
Fix skin component toolbox not working correctly for ruleset matching

Until now, the only usage of ruleset layers was where there is both a
ruleset specific and non-ruleset-specific layer present. The matching
code was making assumptions about this.

As I tried to add a new playfield layer which breaks this assumption,
non-ruleset-specifc components were not being displayed in the toolbox.
This turned out to be due to a `target` of `null` being provided due to
the weird `getTarget` matching (that happened to *just* do what we
wanted previously due to the equals implementation, but only because
there was a container without the ruleset present in the available
targets).

I've changed this to be a more appropriate lookup method, where the
target for dependency sourcing is provided separately from the ruleset
filter.

alexpado commented on osu.Game/Overlays/SkinEditor/SkinComponentToolbox.cs:
```diff
@@ -23,14 +24,22 @@ public partial class SkinComponentToolbox : EditorSidebarSection
+        /// </summary>
+        /// <param name="target">The target. This is mainly used as a dependency source to find candidate components.</param>
+        /// <param name="ruleset">A ruleset to filter components by. If null, only components which are not ruleset-specific will be included.</param>
+        public SkinComponentToolbox(SkinComponentsContainer target, RulesetInfo? ruleset)
+            : base(ruleset == null ? SkinEditorStrings.Components : LocalisableString.Interpolate($"{SkinEditorStrings.Components} ({ruleset.Name})"))
```
Unfortunatly, this breaks the test at `TestSceneSkinEditorComponentsList.TestToggleEditor()`

bdach replied:
Fixed in aca8310cd1df87a131e1e1dfc7d0bc047101c578.

bdach committed:
Fix non-compiling test

To be fair, currently the test is a bit pointless (as it has no reason
to be a `SkinnableTestScene`, it gains precisely nothing from it - all
that is shown there is some generic components on song select). But that
is no worse then `master`, so look away for now.

bdach committed:
Merge branch 'master' into playfield-skin-layer

bdach committed:
Fix tests crashing due to `HUDOverlay` not finding `DrawableRuleset` in `Update()`

bdach commented:
So I've fixed a few more test failures, but the last ones in `SkinDeserialisationTest` I'm not super sure what to do with. They're failing because the ZIP archives with test skins don't have `Playfield.json` inside, thus dying at

https://github.com/ppy/osu/blob/c21167e2ab3b6fd05e54d69537caeca13346c5db/osu.Game/IO/Archives/ZipArchiveReader.cs#L32-L34

But that is curious because skins from before this change continue to work, and I'm pretty sure this failure is _specific to reading skins directly from archives only_. The actual game will use a `DllResourceStore` or a `NativeStorage` underneath, that will just graciously return a null stream, and `Skin` will continue on its merry way.

The `FileNotFoundException` throw in `ZipArchiveReader` is 2016 code, so it's a bit scary to remove it... do we just add blank files to all of the zipped skins or what?

peppy committed:
Return `null` intead of `FileNotFoundException` on missing file in `ZipArchiveReader`

peppy commented:
I'm willing to make a lunge and nuke that exception. It doesn't seem to be used anywhere and definitely doesn't match expectations.

bdach committed:
Merge branch 'master' into playfield-skin-layer

bdach committed:
Fix playfield skin layer not rotating correctly with barrel roll mod
